sdks/java/io/couchbase/src/main/java/org/apache/beam/sdk/io/couchbase/CouchbaseIO.java

```diff
@@ -0,0 +1,359 @@
+     *
+     * @param password password
+     * @return a {@link PTransform} reading data from Couchbase
+     */
+    public Read withPassword(String password) {
+      checkArgument(password != null, "password can not be null");
+      return builder().setPassword(password).build();
+    }
+
+    @Override
+    public PCollection<JsonDocument> expand(PBegin input) {
+      checkArgument((hosts() != null), "WithHosts()is required");
+      checkArgument(bucket() != null, "withBucket() is required");
+
+      CouchbaseSource source = new CouchbaseSource(this);
+      PCollection<JsonDocument> result = input.apply(org.apache.beam.sdk.io.Read.from(source));
+      // Disconnect the client from Couchbase
+      source.close();
+      return result;
+    }
+  }
+
+  @VisibleForTesting
+  static class CouchbaseSource extends BoundedSource<JsonDocument> {
+
+    private final Read spec;
+    private int itemCount;
+    private final int lowerBound; // Lower bound of key range (included)
+    private final int upperBound; // Upper bound of key range (excluded)
+    private Cluster cluster;
+    private Bucket bucket;
+
+    CouchbaseSource(Read spec) {
+      this(spec, null, null, 0, 0);
+    }
+
+    CouchbaseSource(Read spec, Cluster cluster, Bucket bucket, int lb, int ub) {
+      this.spec = spec;
+      this.cluster = cluster;
+      this.bucket = bucket;
+      this.lowerBound = lb;
+      this.upperBound = ub;
+    }
+
+    private Bucket getBucket() {
+      if (cluster == null) {
+        DefaultCouchbaseEnvironment.Builder builder = DefaultCouchbaseEnvironment.builder();
+        if (spec.httpPort() != null) {
+          builder.bootstrapHttpDirectPort(spec.httpPort());
+        }
+        if (spec.carrierPort() != null) {
+          builder.bootstrapCarrierDirectPort(spec.carrierPort());
+        }
+        cluster = CouchbaseCluster.create(builder.build(), spec.hosts());
+      }
+      if (bucket == null) {
+        // For Couchbase Server, in the previous version than 5.0, the passwordless bucket can be
+        // supported.
+        // But after version 5.0, the newly created user should have a username equal to bucket name
+        // and a password.
+        // For more information, please go to
+        // https://docs.couchbase.com/java-sdk/2.7/sdk-authentication-overview.html#legacy-connection-code
+        bucket =
+            spec.password() == null
+                ? cluster.openBucket(spec.bucket())
+                : cluster.openBucket(spec.bucket(), spec.password());
+      }
+      return bucket;
+    }
+
+    @Override
+    public Coder<JsonDocument> getOutputCoder() {
+      return SerializableCoder.of(JsonDocument.class);
+    }
+
+    @Override
+    public List<? extends BoundedSource<JsonDocument>> split(
+        long desiredBundleSize, PipelineOptions options) {
+      // If the desiredBundleSize equals to 0, it means that there will be only one bundle of data
+      // to be read.
+      int totalBundle =
```

Comment:
Since itemCount/estimatedSizeBytes could be null you may need to do `long estimatedSizeBytes = getEstimatedSizeBytes(options);` before.